runtime/major_gc.c

```diff
@@ -1017,6 +1021,13 @@ static void cycle_all_domains_callback(struct domain* domain, void* unused,
 
   caml_ev_begin("major_gc/roots");
   caml_do_roots (&caml_darken, NULL, domain, 0);
+  if (atomic_load_acq(&domain_finished_global_roots) == 0){
```

Comment:
I'm not a fan of having both the lock and the atomic variable acting as synchronization. Can we just use atomic `domain_finished_global_roots` as the only synchronization primitive?

Reply:
Agree that having both lock and atomic variable adds more complexity. The issue I think we might have with using just the atomic variable is this:
* Domain 1 goes into `caml_scan_global_roots` and is not done yet.
* Domain 2 reaches the same point. Seeing that the atomic variable is still 0, it starts scanning the global roots, meaning we could end up with multiple domains scanning global roots.

Reply:
You might need to alter the meaning of the atomic. One way to do it might be to have `domain_global_roots_started` as an enum:
- `0` for `work_unstarted`
- `1` for `work_started`

Then a domain can CAS to claim the work when it is `work_unstarted`, with everybody else skipping it if the work has been started.  

The variable can be initialized in the `caml_global_barrier_is_final` block by the last entrant to the section.

Reply:
As it happens, the current code isn't enforcing only one domain does the work. Consider this sequence:
 - domain A sees `domain_finished_global_roots` is `0` but then suspends
 - domain B sees `domain_finished_global_roots` is `0` takes the lock, does all the work and then releases the lock
 - domain A goes to take the lock, gets it and then does all the work again

Reply:
Yes, got it. This is a great suggestion, thanks for the help! I've added a patch with a similar approach.